### ISSUE TITLE
[Feature Request] Deeply type-check `set` and `frozenset` containers

### ISSUE THREAD

**Issue by langfield** [beartype/beartype#167]
I've only just noticed that the `README` says these are only shallowly-checked. I'm very surprised, as these seem like bread-and-butter data structures. Is it accurate to say that whenever we write `x: Set[T]` (for some specific `T`) it really just checks against `x: Set[Any]`?

What's the roadmap look like for getting this implemented/merged? I'd love to look into how much work a PR for this would be.

**Comment by TeamSpen210** [comment]
Checking them isn't the hard part, the hard part is doing it in amortized `O(1)` time. The way beartype checks an arbitrary data structure without checking time being dependent on the size of the structure is to randomly pick a single element to check each time. That way repeated executions will gradually cover the whole space. But for sets and dicts we cannot sample a random element, at least without iterating through everything before that. Issue #7 has more info on all this. 

I wonder if there's a way to do this via `ctypes` abuse - it's going to be implementation-specific regardless, may as well use that.

**Comment by leycec** [comment]
@TeamSpen210 has the right of it, as always – but @langfield also sorta has the right of it, too.

@beartype absolutely *should* be deeply type-checking at least the first item of standard non-sequence containers like `dict`, `set`, and `frozenset`. That's the bear minimum that @beartype should be performing here.

## The Future Beckons with a Bony Finger

Once we implement that, there are two alternative schools of thought on how to efficiently deeply type-check all of the *other* items of containers like that:

* @TeamSpen210's sagely advice to maliciously abuse `ctypes` would just be the bear's knees. Clearly, CPython itself has some sort of random access to hashable buckets. If CPython does it, so can we if we break enough things with a vicious enough boom stick. That's how hash maps work – which means that's how CPython's `dict`, `set`, and `frozenset` ultimately work, too.
* Alternately, @beartype could internally cache strong references to iterators into previously type-checked `dict`, `set`, and `frozenset` containers. On the next type-check of the same container, @beartype then iterates the corresponding iterator that it previously cached to sequentially access the next item of that container. Clearly, this is dangerous. The naive implementation of this strategy would prevent garbage collection of otherwise dead containers previously type-checked by @beartype. Admittedly, that's horrible. That's why we came up with a crazy two-year-old plan to circumvent that. That's also why we never did that. We could, though – if we're obsessively careful about routinely evicting strong references to iterators into otherwise dead containers from @beartype's internal cache.

## Who Will Heed the Call of the Future?

My internal roadmap that only exists in my head has deep type-checking currently deferred to `beartype 0.13.0`. Specifically:

1. `beartype 0.11.0`, to be released Please-Gods-Let-It-Be-This-Friday, is basically *fait accompli.* Let's pretend I've already pushed that out.
2. `beartype 0.12.0`, to be released Please-Gods-Let-It-Be-Sometime-in-November, implements import hooks and [PEP 526 support](https://peps.python.org/pep-0526/) (i.e., automatic statement-level runtime type-checking of variable assignment annotations) via the nascent `beartype.claw` submodule. This is my next highest priority, because the @beartype decorator itself is now the greatest barrier to @beartype adoption. [~~FaceBook~~ Meta themselves want this](https://github.com/beartype/beartype/issues/43#issuecomment-890361379). What the Zuck wants, the Zuck gets.
3. `beartype 0.13.0`, to be released Please-Gods-Let-It-Be-Sometime-in-February, implements full-blown deep type-checking support for everything standard. So, all core container types and almost everything in `collections.abc`.

If you have a spare lifetime on hand, @langfield, you can lend a helping paw. A PR would be the absolute honey-pot! If you'd like to see the fearsome foe that you'll up against, take a gander at [@beartype's dynamic code generator](https://github.com/beartype/beartype/blob/dd95a2df62dfb507678c406c07f2aa99772eab8f/beartype/_check/expr/exprmake.py#L154). It's over-optimized for extreme speed *without* sacrificing extreme accuracy, which is why that submodule offers a unique window into Hell incarnate. That's where deep type-checking for these containers will need to be added, I'm afraid. And I am, indeed, afraid. Look at that code and weep like a mewling kitten without milk, @langfield! :crying_cat_face: 

Technically, all of this is a duplicate of #53. Pragmatically, more public shaming of my slothful implementation strategy is always a good thing. Let's leave this open to force my unwitting hand.

**Comment by leycec** [comment]
**You were right about everything,** @langfield. I take everything back. `beartype.claw` can wait. @beartype's core mandate is deep type-checking. To the extent that we're not doing that, we're failing.

Here's the new business plan – Gantt charts not included:

* `beartype 0.12.0` to be released hopefully this Friday focuses on **beartype configuration.** *Irrelevant! Next.*
* `beartype 0.13.0` to be released hopefully sometime in November will focus on **deep type-checking.** We'll start with sets and gradually grind our way out to dictionaries. With much luck, intestinal fortitude, and grizzled beards, we will accomplish all our goals.

Thanks for setting me straight, @langfield. You always being right is disconcerting, yet comforting. :smile:

**Comment by langfield** [comment]
@langfield Haha, I don't want to disconcert anyone! Why the change of heart?

**Comment by leycec** [comment]
> Why the change of heart?

Because you're right about everything, of course. That's always the reason.

In truth, it's partially in-house APIs at Ionovate (i.e., a slow-burn tech startup quietly founded by my wife and I). Now that @beartype transparently handles classes and thus dataclasses, type-checking sets and dictionaries in dataclass attributes is a thing we *really* want to happen.

In another truth, it's also partially **[Plum 2](https://github.com/wesselb/plum/issues/68)**: a next-generation multiple dispatch framework dispatching on PEP-compliant type hints. Plum would like to replace its ad-hoc introspection of type hints with `beartype.door.TypeHint`. We would all like that, too.

Sadly, @beartype's depressing lack of deep type-checking would prevent Plum 2 from trivially dispatching on sets and dictionaries: e.g.,

```python
from plum import dispatch

# Beartype sees no problem here!
@dispatch
def f(x: dict[str]) -> str:
    return 'Dictionary of strings right here, bro.'

# Okay. Beartype now sees a problem.
@dispatch
def f(x: dict[int]) -> str:
    return 'Dictionary of integers comin' right up.'
```

@beartype naively treats `dict[str]` and `dict[int]` as semantically equivalent. Dispatching blows up with an unresolved ambiguity and the bear hangs its grizzled muzzle in sadness.

> Apologize for ghosting this issue.

Oh – no worries whatsoever! @beartype internals scare ~~even~~ *especially* me.  :smirk:

**Comment by langfield** [comment]
@leycec What is `dict[str]` or `dict[T]`? Are the keys implicitly strings or something?

**Comment by leycec** [comment]
**Agh!** I am being even more confusing than usual. Yes, `dict[str]` is literally the PEP 585 type hint `dict[str]`. By dispatching on type hints, [Plum](https://github.com/wesselb/plum) is (basically) building out a runtime implementation of the [PEP 484 `@typing.overload` decorator](https://docs.python.org/3/library/typing.html#typing.overload) that actually does something useful: multiple dispatch, yo!

Admittedly, I've never actually *used* multiple dispatch before. I just blindly throw around CompSci jargon like I know what I'm talking about. But now that I know about this Plum thing, I'll be multiple-dispatching everything everywhere regardless of whether I need to. Object-oriented inheritance is so *passé.* :wink:

**Comment by langfield** [comment]
Any updates on when we'll be able to typecheck dictionaries and sets?

**Comment by leycec** [comment]
**GAAAH!!!** The sweat on my brow is now even sweatier. @langfield asks the hard-hitting questions. This... is actually a great question. I've thrown 140% of my volunteer (mostly unpaid) open-source time this spring and summer at our [upcoming `beartype.claw` API implementing import hooks for @beartype](https://github.com/beartype/beartype/issues/43). Without import hooks, @beartype is mostly useless for most use cases (e.g., existing medium- to large-scale codebases). `beartype.claw` has my absolute attention until tested as fully working without exploding everyone's code. That should happen by August 1st.

But then there's a glut of [*other* unrelated issues concerning forward references](https://github.com/beartype/beartype/issues). Since @beartype works without complaint for our team, I'm considering quietly ignoring those issues for the moment. If so, that would free up considerably more time to attend to this long-standing issue.

In short... *I dunno.* I dunno, man! @beartype is mostly unfunded, mostly just worked on by me in my scant free time, and pulled incoherently in a bazillion different fascinating directions. This feature request *will* happen, because it *must* happen. But promising a hard deliverable date leaves my stomach queasy, uneasy, and rumbling with acid reflux. Let us instead simply hope this happens by 2024. My hands are waving frantically.

If it doesn't, I will temporarily stop playing video games until it does! `<-- solemn promise`

**Comment by langfield** [comment]
Hahah I respect your acknowledgement of uncertainty! Yes definitely it seems like a lot of people want import hooks.

Also very impressed this is a spare-time project for you, I was under the impression you were full-time on @beartype, judging by the activity levels. 😲

**Comment by leycec** [comment]
> I was under the impression you were full-time on @beartype, judging by the activity levels.

*Awwww!* Sweet soothing words such as these are like milk and honey to my ears, if ears could eat milk and/or honey. I probably shouldn't have envisioned that.

I shall now reveal the stunning story of how @beartype was built with salty tears, rictus grins, and empty promises. Gather round! Here goes. I substitute the one hour a day of useless free time I would've otherwise just squandered on joy, fun, and fond memories for open-source enthusiasm. That's it. That's the story. :face_exhaling: 

Oddly, it just kinda works. One hour a day may not sound like much. <sup>*...okay, it's not.*</sup> But I contemplate and assiduously plan all day in my head the exact code I'm going to excrete later that night. It then pours out fully formed like a vestigial code baby during that one hour. This disgusting metaphor is accurate.

The average Canadian experience is markedly different. Purportedly, [nine out of ten Canadians ~~squander~~ *invest* three hours a day on television and digital streaming services as of last year](https://mediaincanada.com/2022/02/11/canadians-spend-21-hours-a-week-watching-tv-and-video-content). My response to these shocking allegations:

> "Where can I find these proverbial three hours a day? I don't even have a television! Maybe that's the problem. I looked hard but all I found was dirty cats covered in ticks sleeping on my bed."

**Comment by posita** [comment]
> ... It then pours out fully formed like a vestigial code baby during that one hour. This disgusting metaphor is accurate. ...

It is taking all my mental strength to resist the temptation to discover at what level of detail this metaphor breaks down.

**Comment by leycec** [comment]
It's... it's *beautiful*!

![@beartype baby](https://media.tenor.com/UHRXCs02Ge4AAAAd/jabba-jabba-the-hutt.gif)

Once you see the @beartype code baby, you cannot unsee the @beartype code baby.

**Comment by leycec** [comment]
**Partially resolved by 9f227bd.** Our upcoming @beartype 0.18.0 release now **deeply type-checks all dictionary type hints** in `O(1)` time.** To guarantee this extreme time constraint, @beartype currently *only* type-checks the first key-value pair of dictionaries. The price of speed is high. Still, this is a dramatic improvement over @beartype's prior approach:

> "I didn't see nuffin'."

There are a *shocking* number of ways to annotate dictionaries. @beartype now supports all of them except `typing.TypeDict`, which is kinda out of scope for this feature request. Also, I'm lazy. Specifically, @beartype now supports all: 

* `dict[..., ...]` type hints.
* `collections.defaultdict[..., ...]` type hints.
* `collections.abc.Mapping[..., ...]` type hints.
* `collections.abc.MutableMapping[..., ...]` type hints.
* `collections.abc.OrderedDict[..., ...]` type hints.
* `typing.Dict[..., ...]` type hints.
* `typing.Mapping[..., ...]` type hints.
* `typing.MutableMapping[..., ...]` type hints.
* `typing.OrderedDict[..., ...]` type hints.

Let's kick this into high octane. Consider this minimal example integrating `@beartype` **+** nested dictionary type hints:

```python
from beartype import beartype
from collections.abc import Mapping, MutableMapping

@beartype
def go_bear(bear_bros_for_life: dict[
    int, Mapping[str, MutableMapping[bytes, bool]]]) -> None:
    print(bear_bros_for_life)

# This passes. A beautiful dictionary brings a tear to the eye.
go_bear({
    1: {
        'Beautiful bird;': {
            b'thou voyagest to thine home,': False,
        },
    },
})

# This fails! A horrible dictionary brings your app crashing to the ground.
go_bear({
    1: {
        'With thine,': {
            b'and welcome thy return with eyes': 1,
        },
    },
})
```

...which prints:

```python
{1: {'Beautiful bird;': {b'thou voyagest to thine home,': False}}}
Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 21, in <module>
    go_bear({
  File "<@beartype(__main__.go_bear) at 0x7f75633de2a0>", line 91, in go_bear
beartype.roar.BeartypeCallHintParamViolation: Function __main__.go_bear() parameter
bear_bros_for_life={1: {'With thine,': {b'and welcome thy return with eyes': 1}}}
violates type hint dict[int, collections.abc.Mapping[str,
collections.abc.MutableMapping[bytes, bool]]], as dict key int 1 value dict key str
'With thine,' value dict key bytes b'and welcome thy return with eyes' value int 1
not instance of bool.
```

@beartype thus generates human-readable violation exceptions describing the exact path to the invalid key-value pair. Python QA will never be the same again. :muscle: :bear:

**Comment by leycec** [comment]
**Finally fully resolved by 38cfc61ce64c.** @beartype now deeply type-checks **reiterables** (i.e., collections satisfying the `collections.abc.Collection` protocol with guaranteed `O(1)` read-only access to *only* the first collection item). This includes *all* containers annotated by one or more of the following PEP 484- or 585-compliant type hints:

* `frozenset[...]`.
* `set[...]`.
* `collections.deque[...]`.
* `collections.abc.Collection[...]`.
* `collections.abc.KeysView[...]`.
* `collections.abc.MutableSet[...]`.
* `collections.abc.Set[...]`.
* `collections.abc.ValuesView[...]`.
* `typing.AbstractSet[...]`.
* `typing.Collection[...]`.
* `typing.Deque[...]`.
* `typing.FrozenSet[...]`.
* `typing.KeysView[...]`.
* `typing.MutableSet[...]`.
* `typing.Set[...]`.
* `typing.ValuesView[...]`.

The end of deep `O(1)` type-checking is now in sight, bear peeps. Testify! @beartype now deeply type-checks *almost* all possible type hints in `O(1)` time. Obvious exceptions include probably NP-complete shadow madness like:

* **Callable type hints** (i.e., `collections.abc.Callable[...]` and `typing.Callable[...]` type hints).
* **Type variables** (i.e., `typing.TypeVar(...)` type hints).

This is why @leycec will now sleep like bald sleeping beauty for a hundred years. It sounds like bliss!

![](https://media1.tenor.com/m/BzbYuCYSwAoAAAAC/polar-bear-spring.gif)
<sup>*pretty sure that bear ain't movin'*</sup>
